src/gui/qgsqueryresultwidget.cpp

```diff
@@ -0,0 +1,493 @@
+  }
+}
+
+void QgsQueryResultWidget::updateButtons()
+{
+  mFilterLineEdit->setEnabled( mFirstRowFetched );
+  mFilterToolButton->setEnabled( mFirstRowFetched );
+  mExecuteButton->setEnabled( ! mSqlEditor->text().isEmpty() );
+  mLoadAsNewLayerGroupBox->setVisible( mConnection && mConnection->capabilities().testFlag( QgsAbstractDatabaseProviderConnection::Capability::SqlLayers ) );
+  mLoadAsNewLayerGroupBox->setEnabled(
+    mSqlErrorMessage.isEmpty() &&
+    mFirstRowFetched
+  );
+}
+
+void QgsQueryResultWidget::updateSqlLayerColumns( )
+{
+  // Precondition
+  Q_ASSERT( mModel );
+
+  mFilterToolButton->setEnabled( true );
+  mFilterLineEdit->setEnabled( true );
+  mPkColumnsComboBox->clear();
+  mGeometryColumnComboBox->clear();
+  const bool hasPkInformation { ! mSqlVectorLayerOptions.primaryKeyColumns.isEmpty() };
+  const bool hasGeomColInformation { ! mSqlVectorLayerOptions.geometryColumn.isEmpty() };
+  static const QStringList geomColCandidates { QStringLiteral( "geom" ), QStringLiteral( "geometry" ),  QStringLiteral( "the_geom" ) };
+  const auto constCols { mModel->columns() };
+  for ( const auto &c : constCols )
+  {
+    const bool pkCheckedState = hasPkInformation ? mSqlVectorLayerOptions.primaryKeyColumns.contains( c ) : c.contains( QStringLiteral( "id" ), Qt::CaseSensitivity::CaseInsensitive );
+    // Only check first match
+    mPkColumnsComboBox->addItemWithCheckState( c, pkCheckedState && mPkColumnsComboBox->checkedItems().isEmpty() ? Qt::CheckState::Checked : Qt::CheckState::Unchecked );
+    mGeometryColumnComboBox->addItem( c );
+    if ( ! hasGeomColInformation && geomColCandidates.contains( c, Qt::CaseSensitivity::CaseInsensitive ) )
+    {
+      mGeometryColumnComboBox->setCurrentText( c );
+    }
+  }
+  mPkColumnsCheckBox->setChecked( hasPkInformation );
+  mGeometryColumnCheckBox->setChecked( hasGeomColInformation );
+  if ( hasGeomColInformation )
+  {
+    mGeometryColumnComboBox->setCurrentText( mSqlVectorLayerOptions.geometryColumn );
+  }
+}
+
+void QgsQueryResultWidget::cancelRunningQuery()
+{
+  // Cancel other threads
+  if ( mFeedback )
+  {
+    mFeedback->cancel();
+  }
+
+  // ... and wait
+  if ( mQueryResultWatcher.isRunning() )
+  {
+    mQueryResultWatcher.waitForFinished();
+  }
+}
+
+void QgsQueryResultWidget::cancelApiFetcher()
+{
+  if ( mApiFetcher )
+  {
+    mApiFetcher->stopFetching();
+    mApiFetcherWorkerThread.quit();
+    mApiFetcherWorkerThread.wait();
+  }
+}
+
+void QgsQueryResultWidget::startFetching()
+{
+  if ( ! mWasCanceled )
+  {
+    if ( ! mSqlErrorMessage.isEmpty() )
+    {
+      showError( tr( "SQL error" ), mSqlErrorMessage, true );
+    }
+    else
+    {
+      if ( mQueryResultWatcher.result().rowCount() > 0 )
+      {
+        mStatusLabel->setText( QStringLiteral( "Query executed successfully (%1 rows)" ).arg( QLocale().toString( mQueryResultWatcher.result().rowCount() ) ) );
```

Comment:
Minor thing, but I've always liked it how pgadmin reported the total query execution time. (It's useful for easy testing if changes result in faster queries). Might be something to consider here?

Reply:
Yeah, I like that too. At this stage we can only measure the execution time, not the fetching. I'll see if I can add it.

Reply:
Execution time is perfect anyway ... it's more reflective of the load on the database itself vs network connection/transient network slowdown etc.